### PR TITLE
Declarative function examples

### PR DESCRIPTION
**Pull request by MatthieuDartiailh**
This adds some examples on how to use declarative functions. Some examples are not very inspired but at least they show how things work. I would like to add a fourth one so this is not ready to merge. But feel free to review what is already there.

### CODE REVIEW

File: examples/layout/advanced/override_layout_constraints.enaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+<< autodoc-me >>
+"""
+from itertools import izip_longest
+
+from enaml.core.api import Include
+from enaml.layout.api import align, grid, factory
+from enaml.widgets.api import Window, Container, Field, Label, PushButton
+
+
+def generate_grid():
```
Comment (blink1073) [inline examples/layout/advanced/override_layout_constraints.enaml]: Is this function left over?

Comment (MatthieuDartiailh) [inline examples/layout/advanced/override_layout_constraints.enaml]: Indeed I will remove it when adding the last example.

**Comment by jason-s** [comment]
dumb question: why use a special keyword `func` rather than Python's `def`?

```
func do_thing(param):
  print thing, self.thing2, sbox.value, param
```
vs.

```
def do_thing(self, param):
  print self.thing, self.thing2, self.sbox.value, param
```

**Comment by sccolbert** [comment]
Two reasons:

1) It makes it easier to parse
2) A `func` does not behave like a normal function. It has a magic `self` and dynamic scoping.

File: examples/functions/declare_function.enaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+# The full license is in the file COPYING.txt, distributed with this software.
+#------------------------------------------------------------------------------
+""" An example showing how to declare a declarive function using func
+
+This example demonstrates how one can use the `func` keyword to declare
+a function bound to an enaml declaration and how scoping rules apply.
+
+Note that when declaring a function using the `func` it will always be
+considered a method of the object on which it is defined hence the possibility
+to access to self which is a reference to the main window.
```
Comment (sccolbert) [inline examples/functions/declare_function.enaml]: can you change this to "to access `self` which in this case is a reference to the main window".

File: examples/layout/advanced/override_layout_constraints.enaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    """
+    rows = []
+    widgets = self.visible_widgets()
+    row_iters = (iter(widgets),) * 2
+    rows = list(izip_longest(*row_iters))
+    return [grid(*rows), align('width', *widgets)]
+
+
+enamldef Main(Window):
+    title = 'Factory Helper'
```
Comment (sccolbert) [inline examples/layout/advanced/override_layout_constraints.enaml]: this title doesn't make sense with this example

File: examples/layout/advanced/override_layout_constraints.enaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            rows = list(izip_longest(*row_iters))
+            return [grid(*rows)] + [align('v_center', *row) for row in rows]
+
+        Label:
+            text = 'Name'
+        Field:
+            pass
+        Label:
+            text = 'Surname'
+        Field: c_count:
```
Comment (sccolbert) [inline examples/layout/advanced/override_layout_constraints.enaml]: this id isn't used, should probably be removed

File: examples/layout/advanced/override_layout_constraints.enaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+different from the deafults ones.
+
+<< autodoc-me >>
+"""
+from itertools import izip_longest
+
+from enaml.core.api import Include
+from enaml.layout.api import align, grid, factory
+from enaml.widgets.api import Window, Container, Field, Label, PushButton
+
```
Comment (sccolbert) [inline examples/layout/advanced/override_layout_constraints.enaml]: cleanup unused imports

File: examples/functions/override_function.enaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+Note that when subclassing an enaml object you must use the @d_func decoarator
+to mark a function as declarative and hence overridable from enaml.
+
+<< autodoc-me >>
+"""
+from atom.api import observe
+from enaml.core.declarative import d_func
+from enaml.widgets.api import Window, PushButton, Container
+from enaml.layout.api import hbox, vbox
+
```
Comment (sccolbert) [inline examples/functions/override_function.enaml]: cleanup unused imports

File: examples/functions/declare_function.enaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+Note also that such a function has access to the same scope as any other Enaml
+object allowing for example to access the SpinBox using its id which would not
+be possible using a function (defined using def) declared outside the enamldef
+block.
+
+<< autodoc-me >>
+"""
+from enaml.widgets.api import Window, PushButton, SpinBox, Container
+from enaml.layout.api import hbox, vbox
```
Comment (sccolbert) [inline examples/functions/declare_function.enaml]: cleanup unused imports

**Comment by MatthieuDartiailh** [comment]
Comments addressed and last example added.
This last example is a bit more complex and I don't know if it presents the right approach to achieve what it tries to achieve, but it does work and seems nice to me so I thought it would be nice to share it.

File: examples/functions/observe_model_signal.enaml [hunk 8]
```diff
@@ -0,0 +1,125 @@
+#------------------------------------------------------------------------------
+# Copyright (c) 2013, Nucleic Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#------------------------------------------------------------------------------
+""" An example showing how to use to observe a model signal.
+
+Here the model use a signal to notify listeners about in place change to a
```
Comment (sccolbert) [inline examples/functions/observe_model_signal.enaml]: s/use/uses

File: examples/functions/observe_model_signal.enaml [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+list. This pattern is interestign when a ContainerList would emit too
```
Comment (sccolbert) [inline examples/functions/observe_model_signal.enaml]: s/interestign/interesting

File: examples/functions/observe_model_signal.enaml [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+many signal because notifications cannot be batched. This is a common case
```
Comment (sccolbert) [inline examples/functions/observe_model_signal.enaml]: s/signal/signals

File: examples/functions/observe_model_signal.enaml [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    def delete_value(self, index):
+        """Delete a value.
+
+        """
+        del self.values[index]
+        self.values_changed('deleted')
+
+
+enamldef EditMenu(Menu):
+    """Menu use to edit the content of the list.
```
Comment (sccolbert) [inline examples/functions/observe_model_signal.enaml]: s/use/used

File: examples/functions/observe_model_signal.enaml [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+    The entries depend on the state of the list to avoid proposing absurd
```
Comment (sccolbert) [inline examples/functions/observe_model_signal.enaml]: This line is a bit unclear. Can you rephrase?

File: examples/functions/observe_model_signal.enaml [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+enamldef Main(MainWindow): main:
+
+    attr model = Model()
+    attr _internal << model.values[:]
+
+    Form:
+        func react_signal(kind):
+            main._internal = model.values[:]
+            if kind == 'moved':
+                self.request_relayout()
```
Comment (sccolbert) [inline examples/functions/observe_model_signal.enaml]: is this request layout needed?

**Comment by sccolbert** [comment]
Looks pretty good. A few more comments on the new example.

**Comment by MatthieuDartiailh** [comment]
Comments addressed. I added a comment about the request_relayout. I must say I am not sure to understand why it works but it does. One thing that I do not understand is why the order of the children is respected but the Looper does not update their loop_index. It might be something to fix.

File: examples/functions/observe_model_signal.enaml [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+enamldef Main(MainWindow): main:
+
+    attr model = Model()
+    attr _internal << model.values[:]
+
+    Form:
+        func react_signal(kind):
+            main._internal = model.values[:]
+            if kind == 'moved':
```
Comment (sccolbert) [inline examples/functions/observe_model_signal.enaml]: This whole "if" block can be removed. I just pushed an update which fixes the lack of automatic relayout when children are simply moved instead of added/removed.
https://github.com/nucleic/enaml/commit/e0a886a899deb45bb4c255f8fe400fa5db0ca19c

**Comment by sccolbert** [comment]
Just fixed the bug for the relayout request and made a comment on one last thing to fix before merging.

**Comment by MatthieuDartiailh** [comment]
Last comment addressed !

**Comment by sccolbert** [comment]
Thanks!
